### PR TITLE
Elasticsearch nested aggregations

### PR DESCRIPTION
**Pull request by jremmurd**
# Nested aggregations

The nested type is a specialised version of the object datatype that allows arrays of objects to be indexed in a way that they can be queried independently of each other.

This PR enables support for aggregating by nested documents and adds an example filter implementation to the docs.

For details see 
- https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html

### CODE REVIEW

**Comment by fashxp** [comment]
thix, will need some additional time to review and test this.

File: bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php [hunk 1]
```diff
@@ -1135,6 +1159,65 @@ protected function doLoadGroupByValues()
...
     }
 
+    /**
+     * Deep search for buckets in result aggregations array, as the structure of the result array
+     * may differ dependent on the used aggregations (i.e. date filters, nested aggr, ...)
+     *
+     * @param array $aggregations
+     * @return array
+     */
+    private function searchForBuckets(array $aggregations)
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php]: shouldn't we make these methods protected to make overwriting possible?

File: bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php [hunk 2]
```diff
@@ -1135,6 +1159,65 @@ protected function doLoadGroupByValues()
...
+        return [];
+    }
+
+    /**
+     * Recursively convert aggregation data (sub-aggregations possible)
+     *
+     * @param array $bucket
+     * @return array
+     */
+    private function convertBucketValues(array $bucket){
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php]: shouldn't we make these methods protected to make overwriting possible?

File: doc/Development_Documentation/10_E-Commerce_Framework/07_Filter_Service/03_Elastic_Search.md [hunk 3]
```diff
@@ -1,52 +1,180 @@
...
+            type: 'nested'
+    interpreter_id: AppBundle\Ecommerce\IndexService\Interpreter\MyAttributes
+    interpreter_options:
+        locale: '%%locale%%'
+```
+
+Now you can create a filter for the nested document field, which has to be defined in a nested manner as well:
+
+```php
+class SelectMyAttribute extends \Pimcore\Bundle\EcommerceFrameworkBundle\FilterService\FilterType\ElasticSearch\AbstractFilterType
```
Comment (fashxp) [inline doc/Development_Documentation/10_E-Commerce_Framework/07_Filter_Service/03_Elastic_Search.md]: there is no such thing as \Pimcore\Bundle\EcommerceFrameworkBundle\FilterService\FilterType\ElasticSearch\AbstractFilterType in the framework. 
Would that be needed? Maybe missing in the PR?

Comment (fashxp) [reply doc/Development_Documentation/10_E-Commerce_Framework/07_Filter_Service/03_Elastic_Search.md]: how would the `getFilterFrontend` method look like?

File: bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php (hunk 1 above)
Comment (jremmurd) [reply bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php]: makes sense - resolved

File: bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php (hunk 2 above)
Comment (jremmurd) [reply bundles/EcommerceFrameworkBundle/IndexService/ProductList/ElasticSearch/AbstractElasticSearch.php]: makes sense - resolved

File: doc/Development_Documentation/10_E-Commerce_Framework/07_Filter_Service/03_Elastic_Search.md (hunk 3 above)
Comment (jremmurd) [reply doc/Development_Documentation/10_E-Commerce_Framework/07_Filter_Service/03_Elastic_Search.md]: Thx, fixed the namespace. I also added an example implementation for the getFilterFrontend().

**Comment by fashxp** [comment]
thx very much!

**Pull request merged**